### PR TITLE
Dynamic environments

### PR DESCRIPTION
**Pull request by tobiges**
Adds the ability to create environments with the specified parameters.
Current existing environments are modeled using the new dynamic creation.

### CODE REVIEW

File: cc/env/envs/two_segments.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+@dataclass
+class CartParams:
+    name: str
+    slider_joint_params: JointParams
+    hinge_joint_params: JointParams
+
+
+def generate_body(
+    name: str, slider_joint_params: JointParams, hinge_joint_params: JointParams
+) -> str:
```
Comment (SimiPixel) [inline cc/env/envs/two_segments.py:29]: return type is bytes?

File: cc/env/register.py [hunk 2]
```diff
@@ -1,10 +1,18 @@
...
 from dm_control import mujoco
 from dm_control.rl import control
 
-from .envs import two_segments_v1, two_segments_v2, two_segments_v3
+from .envs import two_segments
+from .envs.two_segments import CartParams, JointParams, SegmentTask
+
+
+def load_physics_wrapper(cart_params: List[CartParams]):
+    def load_physics():
```
Comment (SimiPixel) [inline cc/env/register.py:12]: `_load_physics_two_segments` 
better indicate its private to module

also return type is missing

**Comment by SimiPixel** [comment]
I dont really like the function `register_new_two_segment_env`.

How about the following changes:
```python
@dataclass
class EnvRegisterValue:
    load_physics: Callable[[], mujoco.Physics]
    ...
```
where `load_physics` is now a function
```python
def load_physics(**physics_kwargs) -> mujoco.Physics:
    pass
```
and we change `make_env` to
```python
def make_env(
    id: str,
    time_limit: Optional[float] = None,
    control_timestep: Optional[float] = None,
    single_precision: Optional[bool] = True,
    delay: int = 0,
    task_kwargs = {},
    physics_kwargs={}
)
```
where physics_kwargs are then forwarded to `load_physics`.

Then, we can create a new `EnvRegisterValue` with the function `two_segments.load_physics` and we can simply do
```python
env = make_env("two_segments", physics_kwargs={"cart_params": cart_params})
```

This feels more straightforward. What do you think?

**Comment by tobiges** [comment]
We thought a bit about how to improve the make_env function and it may be better to get rid of the EnvRegister since if you pass the parameters directly to the make_env anyways it does not provide any real value.  
Instead we would propose to pass the load_physics together with the other parameters directly to make_env.
Sample configs for two_segemnts_v1 etc. could still be made accessible so that you don't always have to create your own EnvConfig.

````python
@dataclass
class EnvConfig:
    load_physics: Callable[[], mujoco.Physics]
    Task: control.Task
    time_limit: float
    control_timestep: float = 0.01
````

```python
def make_env(
    env_config: EnvConfig,
    single_precision: Optional[bool] = True,
    delay: int = 0,
    **task_kwargs
)
```

What do you think?

**Comment by SimiPixel** [comment]
I am not a big fan on having `make_env` take some magical object. I think its better if there exists one function `make_env` that you always use and change its input to get a certain `Env` out. And all its input are only strings nothing fancy.

Here, one would have to know that the first argument is this special object which i can import from somewhere and that i can actually modifiy the `Env` that `make_env` creates by giving the `EnvConfig` different arguments. 
It's a two step approach where one would suffice (only think about the API). So, what would you prefer if you didn't understand anything internal to `make_env`. 

Then, you can simply explain `make_env` and its docstring is the only thing you need to understand to do it all.

**Comment by tobiges** [comment]
I am not a big fan of these magical strings either. Also with the change you proposed above, you would still need to supply the cart params.

Imo it is quite simple to create a env with the current logic:
```
from cc.env.sample_envs import TWO_SEGMENT_V1
from cc.env.make_env import make_env

env = make_env(TWO_SEGMENT_V1, random=1)
```

What we could do is expose two function:
`make_env_from_config` which works the way we changed `make_env` currently
`make_env` which takes the sting as it is currently and internally calls `make_env_from_config`

Nevertheless if you prefer to have only the string as an option we can also implement that.

File: cc/utils/multiple_controller_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        self.controllers = controllers
+
+    def step(self, x):
+        actions = [{"obs": {"xpos_of_segment_end": s_obs},
+                    "ref": {"xpos_of_segment_end": s_ref}}
+                   for s_obs, s_ref in zip(x["obs"]["xpos_of_segment_end"], x["ref"]["xpos_of_segment_end"])]
+
+        new_controllers, actions = zip(
+            *[rgf.step(actions[index]) for index, rgf in enumerate(self.controllers)])
+        return MultipleControllerWrapper(*new_controllers), np.asarray(actions).flatten()
```
Comment (SimiPixel) [inline cc/utils/multiple_controller_wrapper.py:19]: Could we change the `np` to `jnp`. It should still work right?

**Review by SimiPixel — APPROVED** [review]
Maybe later you could provide one `recording_multiple_controllers.py` for the example folder as a small and simple entry point.
Overall, i think this is a really nice contribution. Thank you!

**Pull request merged**